### ISSUE TITLE
Utilize the already existing code in rosmsg to get message definition?

### ISSUE THREAD

**Issue by peci1** [StefanFabian/ros_babel_fish#2]
Hi, rosmsg already has the implementation of message definition lookup: https://github.com/ros/ros_comm/blob/d487d5877049d61fe2ebcc6487e9a7a72d9f0346/tools/rosmsg/src/rosmsg/__init__.py#L414 . What about utilizing this method in embedded_python_description_provider instead of mimicking the loop in embedded python?

Anyways, thanks for the package, it looks promising!

**Comment by StefanFabian** [comment]
Hey,

thank you for your kind words!
Regarding your issue, this is actually not the same as you can verify with the following commands:
```Python
import rosmsg
print(rosmsg.get_msg_text("geometry_msgs/Pose"))
```
which should give you the following output:
```
geometry_msgs/Point position
  float64 x
  float64 y
  float64 z
geometry_msgs/Quaternion orientation
  float64 x
  float64 y
  float64 z
  float64 w


```
That's not the representation I need which you can see here:
```
$ rosrun ros_babel_fish message_info geometry_msgs/Pose
[...]
# A representation of pose in free space, composed of position and orientation. 
Point position
Quaternion orientation

================================================================================
MSG: geometry_msgs/Point
# This contains the position of a point in free space
float64 x
float64 y
float64 z

================================================================================
MSG: geometry_msgs/Quaternion
# This represents an orientation in free space in quaternion form.

float64 x
float64 y
float64 z
float64 w

```

There may be a method that does what I need that I have missed but unfortunately `get_message_text` isn't that method.
In any case, the embedded python provider is more of a relict that can optionally be used in case I've made a mistake in the integrated description provider which is the provider that is used by default and doesn't use any embedded Python which I consider quite the dirty hack in a library.
For that reason, the embedded python provider is not included by default and can only be enabled using a CMake option.
I'm considering removing it entirely in a future version but the code is there so there was no harm in including it.
I've also considered an external python provider that would use interprocess communication to communicate the message definition but I couldn't get it to work 100% reliably and don't have time to work on that since it's not really important and apart from one part that's still missing the integrated provider works without any issue for me (and the tests).

**TL;DR**
You don't have to care about the embedded python description provider at all unless you find a case where it works and the integrated fails in which case you should create an issue and can use the python version in the meantime.
By default, this library doesn't use python and does not depend on it. You have to build it from source if you want that functionality.

**Comment by peci1** [comment]
The `get_msg_text` method has a `raw` parameter:

    :param raw: if True, include comments and whitespace (default False), ``bool``

My reasoning behind this was that the implementation in rosmsg is the only (and thus I'd say canonical) implementation in the official ROS pacakges. Therefore, I'd consider a better solution "piggybacking" on this implementation than creating another one which may behave differently in some corner cases.

**Comment by StefanFabian** [comment]
I understand your concern but I'd advise you to try the commands you suggest.
Using the raw option:
```Python
import rosmsg
print(rosmsg.get_msg_text("geometry_msgs/Pose", raw=True))
```
will give you the following
```
# A representation of pose in free space, composed of position and orientation. 
Point position
Quaternion orientation

```
which does not include the message's dependencies.

I absolutely agree with that reasoning but it doesn't apply here. I'm using multiple commands but they are all public API and to the best of my knowledge there is no simpler method of obtaining the required information.
What the python provider does is basically run the following python code:
```Python
spec = genmsg.load_msg_by_type(context, name, msg_search_paths)
genmsg.load_depends(context, spec, msg_search_paths)
md5 = genmsg.compute_md5(context, spec)
full_text = genmsg.compute_full_text(context, spec)
```
where `name` is the full name of the message, e.g., "geometry_msgs/Pose".
The retrieval for service descriptions is analogous.
As you can see that's public API and not "piggybacking" on any implementation.

I also don't get why you're so focused on the python description provider. You should **NOT** use it. Only if the native one doesn't work and you need something until I have fixed the native implementation (that is a hypothetical scenario with a probability close to zero).
The native one is implemented just like the official python version and should behave the same unless they would change the official version which they most likely won't because it would cause havoc and disaster for all existing workspaces.

**Comment by peci1** [comment]
Oh, I'm sorry, I somehow overlooked the integrated provider. And I was also not aware that the rosmsg raw option actually turns off recursive field expansion. Then it's not really useful as you say.

Looking at the integrated provider, it has about 350 lines of C++ code. Can you be sure there is no bug in the whole file? Because if there is, then the provider would behave different from the python/roscore behavior and hard to debug errors will pop up... That's why i'd prefer piggybacking on rosmsg or other python interfaces rather than reimplementing the algorithms in C++. But I understand the embedded Python solution is also not really a good one. I think the best solution would be if rosmsg directly generated C++ bindings using e.g. SWIG (or, even better, if they reimplemented it in C++ and then made the Python interface via SWIG instead). But that's probably something that won't happen... 

So I think we can conclude that there is (and probably won't be) a nice solution to this problem?

Thanks for your time.

**Comment by StefanFabian** [comment]
You can never be sure about the absence of bugs. That's impossible.
However, the integrated description provider has close to 100% line coverage with multiple test messages of varying complexity. Hence, it is somewhat unlikely that there's a bug in there that the tests wouldn't cover.
If there was a bug, the error would be pretty easy to spot because the md5 sums wouldn't match.
I disagree that there's no nice solution. Reimplementing the algorithm in C++ is as nice as it gets (conditioned on the fact that they can't change that algorithm because it would break everything).
There is no perfect solution but perfect solutions generally don't exist with very few exceptions.

**Comment by peci1** [comment]
I agree it's probable that your reimplementation is correct. 

What I wanted to point out as "nice solution" is the "single implementation" paradigm, which means having the implementation only at a single place and providing APIs for other languages. That's generally much better for maintenance. It might for example happen that ROS will extend the message searching functionality in future versions (in a backwards compatible way), and if you don't notice this change, this package will break for messages utilizing the new functionality even though it is fully tested. This is the kind of stuff you'd prevent by having just a single implementation.

**Comment by StefanFabian** [comment]
True, that would be an even nicer solution but unfortunately it's quite unlikely that this will happen unless you want to put in the work and make a PR.

**Comment by peci1** [comment]
I agree. And I don't believe this could change in ROS1.
